### PR TITLE
fix: incorrect physical types of character array

### LINKED ISSUE #4404: [Compile LAPACK] declaring arrays in subroutines
I believe I found a problem with the declaration of arrays in subroutines when the length for the array is a parameter as well.
I have the following example program derived from `blas/src/xerbla_array.f` [here](https://github.com/Reference-LAPACK/lapack):
```
      PROGRAM EXAMPLE_ERROR
            
            INTEGER A
            CHARACTER B(1)
            
            A = 1
            CALL ERROR_SUBROUTINE(B,A)

      END PROGRAM


      SUBROUTINE ERROR_SUBROUTINE(B, A)
      
            INTEGER A 
            CHARACTER B(A)
            
            B = 'X'
            PRINT *, B

            RETURN

      END
```

The output of `gfortran example.f` is `X`, whereas `lfortran --fixed-form --show-llvm example.f` is:
```
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/kerimbirgi/dev/lfortran/src/bin/lfortran.cpp", line 2478
    LCompilers::print_stack_on_segfault();
  File "/Users/kerimbirgi/dev/lfortran/src/bin/lfortran.cpp", line 2362
    return emit_llvm(arg_file, lfortran_pass_manager,
  File "/Users/kerimbirgi/dev/lfortran/src/bin/lfortran.cpp", line 852
    = fe.get_llvm(input, lm, pass_manager, diagnostics);
  File "/Users/kerimbirgi/dev/lfortran/src/lfortran/fortran_evaluator.cpp", line 295
    Result<std::unique_ptr<LLVMModule>> res = get_llvm2(code, lm, pass_manager, diagnostics);
  File "/Users/kerimbirgi/dev/lfortran/src/lfortran/fortran_evaluator.cpp", line 317
    diagnostics, lm.files.back().in_filename);
  File "/Users/kerimbirgi/dev/lfortran/src/lfortran/fortran_evaluator.cpp", line 363
    compiler_options, run_fn, infile);
  File "/Users/kerimbirgi/dev/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 9973
    v.visit_asr((ASR::asr_t&)asr);
  File "../libasr/asr.h", line 5154
  File "../libasr/asr.h", line 5130
  File "../libasr/asr.h", line 5155
  File "../libasr/asr.h", line 4861
  File "/Users/kerimbirgi/dev/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 953
    visit_Function(*ASR::down_cast<ASR::Function_t>(item.second));
  File "/Users/kerimbirgi/dev/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3858
    llvm_goto_targets.clear();
  File "/Users/kerimbirgi/dev/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 3899
    llvm::FunctionType* function_type = llvm_utils->get_function_type(x, module.get());
  File "/Users/kerimbirgi/dev/lfortran/src/libasr/codegen/llvm_utils.cpp", line 1133
    std::vector<llvm::Type*> args = convert_args(x, module);
  File "/Users/kerimbirgi/dev/lfortran/src/libasr/codegen/llvm_utils.cpp", line 911
    type_original = get_arg_type_from_ttype_t(arg->m_type,
  File "/Users/kerimbirgi/dev/lfortran/src/libasr/codegen/llvm_utils.cpp", line 668
    LCOMPILERS_ASSERT(false);
AssertFailed: false
```
Both `--show-ast` and `--show-asr` produce outputs therefore I assume the error happens during the ASR to LLVM translation. Any help or guidance is appreciated.

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #4404.

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -3437,7 +3437,9 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
             type = ASRUtils::TYPE(ASR::make_Character_t(al, loc, a_kind, a_len, len_expr));
             type = ASRUtils::make_Array_t_util(
                 al, loc, type, dims.p, dims.size(), abi, is_argument,
-                dims.size() > 0 ? ASR::array_physical_typeType::CharacterArraySinglePointer : ASR::array_physical_typeType::DescriptorArray,
+                dims.size() > 0 && abi == ASR::abiType::BindC ? ASR::array_physical_typeType::CharacterArraySinglePointer :
+                                ASRUtils::is_fixed_size_array(dims.p, dims.n) ? ASR::array_physical_typeType::FixedSizeArray :
+                                ASR::array_physical_typeType::DescriptorArray,
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:3442]: This is done based on https://github.com/lfortran/lfortran/pull/2929#issuecomment-1845897861

File: src/libasr/asr_utils.cpp [hunk 2]
```diff
@@ -1583,7 +1583,7 @@ void make_ArrayBroadcast_t_util(Allocator& al, const Location& loc,
     dims.push_back(al, dim);
     ASR::ttype_t* dest_shape_type = ASRUtils::TYPE(ASR::make_Array_t(al, loc,
         ASRUtils::TYPE(ASR::make_Integer_t(al, loc, 4)), dims.p, dims.size(),
-        is_value_character_array ? ASR::array_physical_typeType::CharacterArraySinglePointer: ASR::array_physical_typeType::FixedSizeArray));
+        is_value_character_array && !is_value_constant(expr2) ? ASR::array_physical_typeType::CharacterArraySinglePointer: ASR::array_physical_typeType::FixedSizeArray));
```
Comment (Pranavchiku) [inline src/libasr/asr_utils.cpp:1586]: More specifically it should be exact same as what we did in `ast_common_visitor`

File: src/libasr/asr_utils.cpp [hunk 3]
```diff
@@ -1610,7 +1610,7 @@ void make_ArrayBroadcast_t_util(Allocator& al, const Location& loc,
             ASRUtils::get_fixed_size_of_array(expr1_mdims, expr1_ndims) <= 256 ) {
             ASR::ttype_t* value_type = ASRUtils::TYPE(ASR::make_Array_t(al, loc,
                 ASRUtils::type_get_past_array(ASRUtils::expr_type(expr2)), dims.p, dims.size(),
-                is_value_character_array ? ASR::array_physical_typeType::CharacterArraySinglePointer: ASR::array_physical_typeType::FixedSizeArray));
+                is_value_character_array && !ASRUtils::is_value_constant(expr2) ? ASR::array_physical_typeType::CharacterArraySinglePointer: ASR::array_physical_typeType::FixedSizeArray));
```
Comment (Pranavchiku) [inline src/libasr/asr_utils.cpp:1613]: this as well.

**Comment by Pranavchiku** [comment]
I think WASM failure is not related to this PR.

**Comment by kmr-srbh** [comment]
> I think WASM failure is not related to this PR.

This issue occurs in all the subsequent PRs also. I think it is related to the CI itself.

**Comment by certik** [comment]
See https://github.com/lfortran/lfortran/issues/4431 for the WASM failure.

File: integration_tests/character_06.f90 [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+subroutine error_subroutine(b, a)
+integer :: a, i
+character :: b(a)
+b = 'x'
+print *, b
+do i = 1, a
+    if (b(i) /= 'x') error stop
+end do 
+return
+end
```
Comment (certik) [inline integration_tests/character_06.f90:17]: Let's put this either into a module or inside the program. When you write it like this, GFortran will use implicit interface (I think) or implicit typing. In LFortran we use an explicit interface here, but the standard Fortran way to write this is to give the compiler access to the function, so it needs to be in a module or inside the main program.

**Review by certik — APPROVED** [review]
I think it's fine. If this fixes it, I think we can merge it. The changes seem ok at the ASR level. If we need to later rework the strings more, we can.

**Pull request merged**